### PR TITLE
Add custom provider

### PR DESCRIPTION
**Pull request by evanli18**

### CODE REVIEW

**Comment by msftgits** [comment]
[CLA assistant check](https://cla.opensource.microsoft.com/oam-dev/terraform-controller?pullRequest=215) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/oam-dev/terraform-controller?pullRequest=215) before we can accept your contribution.<br/><br/>:x: evanli18 [sign now](https://cla.opensource.microsoft.com/oam-dev/terraform-controller?pullRequest=215)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/oam-dev/terraform-controller?pullRequest=215) it.</sub>

**Comment by evanli18** [comment]
#203

File: codecov.yml [hunk 1]
```diff
@@ -7,3 +7,9 @@ coverage:
     patch:
       default:
         target: 80%
+        paths:
+          - "!controllers/provider"
+      credentials:
+        paths:
+          - controllers/provider
+        target: 40%
```
Comment (zzxwill) [inline codecov.yml:15]: Let's remove these lines. For the first time of the contribution, it's okay not to meet the target:)

**Comment by zzxwill** [comment]
This implementation is interesting. Have you built your own Terraform docker image https://github.com/oam-dev/terraform-controller/blob/master/chart/values.yaml#L10?

File: codecov.yml (hunk 1 above)
Comment (evanli18) [reply codecov.yml:15]: > This implementation is interesting. Have you built your own Terraform docker image https://github.com/oam-dev/terraform-controller/blob/master/chart/values.yaml#L10?

Not yet. I'm using private registries

Comment (zzxwill) [reply codecov.yml]: But you still need a self-built Terraform docker image?

Comment (zzxwill) [reply codecov.yml]: I hope to have a whole story on how to use the custom provider, and others can use the feature you contributed.

Comment (evanli18) [reply codecov.yml]: I still use the terrain from image provided by OAM.

When terraform is applied, the provider defined in HCl will be pulled.

The custom provider (credentials) is used to solve the problem of private provider authentication. The existing credentials field cannot meet my needs. At the same time, I do not want to disclose my authentication field and other extended fields (business types, e.g: COST_ACCOUNT)

Comment (zzxwill) [reply codecov.yml]: Thanks, I got the point. 

One last question, can you post an example of your Configuration (especially your HCL/.tf) without any sensitive information?

Comment (evanli18) [reply codecov.yml]: Ok, let me add that.

**Review by zzxwill — APPROVED** [review]
Great job!

**Comment by zzxwill** [comment]
@evanli18 Can you contact me by my email zzxwill@gmail.com or my DingTalk bag82xp. I wonder if we could have a conversion about cloud resource provisioning? Thanks.

**Pull request merged**